### DISCUSSION TITLE
integrating vps with react-stream and hattip

### DISCUSSION THREAD

**Discussion by advancedtw** [vikejs/vike#933] [Discussion] [answered]
Hi i'm trying to make react-streaming + hattip + vps work on node and cloudflare.
I have successfully made vps + react-streaming work on cloudflare (wrangler 3)  and node. now I'm trying to integrate hattip following the [example](https://github.com/hattipjs/hattip/tree/main/examples/vavite/vite-plugin-ssr) in the repo but adapting it to react-stream. 
I've got that so far 
```typescript
// <reference types="vite/client" />

import { renderPage } from "vite-plugin-ssr/server";
import { compose, RequestContext } from "@hattip/compose";

export default compose(handler);

async function handler(ctx: RequestContext) {
        const { readable, writable } = new TransformStream();

        const parsedUrl = new URL(ctx.request.url);
        const url = parsedUrl.pathname + parsedUrl.search;
        const userAgent = ctx.request.headers["user-agent"];

        const pageContextInit = { urlOriginal: url, userAgent };
        const pageContext = await renderPage(pageContextInit);
        const { httpResponse } = pageContext;

        if (!httpResponse) return ctx.next();

        httpResponse.pipe(writable);

        return new Response(readable, {
                status: httpResponse.statusCode,
        });
}
```

I ran my code through vavite but I'm not sure if it is needed or if there is a better solution.
the error message I get is 
```
Error: [vite-plugin-ssr@0.4.131][Wrong Usage] pageContext.httpResponse.pipe() can't be used because the onRenderHtml() hook defined by /renderer/+onRenderHtml.tsx > `export default` provides a Node.js Stream Pipe. See https://vite-plugin-ssr.com/stream for more information.
```

that is the error message when I run it locally on node. on Cloudflare the error message is not helpful so I didn't include it

this is what +onRenderHtml.tsx looks like
```typescript
export default onRenderHtml;

import React from "react";
import { renderToStream } from "react-streaming/server";
import { escapeInject } from "vite-plugin-ssr/server";
import { PageLayout } from "./PageLayout";

async function onRenderHtml(pageContext: any) {
        const { Page, pageProps } = pageContext;

        const page = (
                <PageLayout pageContext={pageContext}>
                        <Page {...pageProps} />
                </PageLayout>
        );

        const stream = await renderToStream(page, { userAgent: pageContext.userAgent });

        return escapeInject`
        <!DOCTYPE html>
        <html lang="en">
          <Head>
            <meta name="viewport" content="width=device-width, initial-scale=1" />
          </Head>
          <body>
            <div id="page-view">${stream}</div>
          </body>
        </html>`;
}
```

any idea what is wrong ?

**Comment by brillout** [comment]
> Error: [vite-plugin-ssr@0.4.131][Wrong Usage] pageContext.httpResponse.pipe() can't be used because the onRenderHtml() hook defined by /renderer/+onRenderHtml.tsx > `export default` provides a Node.js Stream Pipe. See https://vite-plugin-ssr.com/stream for more information.

This is misleading. The problem is actually that you're trying to use a Web Stream together with a Node.js Stream. (I'll improve the error message in a moment.) Make both Node.js Streams or Web Streams, but don't mix both.

You use `react-streaming` which provides a Web Stream on Cloudflare and a Node.js Stream on Node.js. That's likely the problem as I believe HatTip always expects a Web Stream. Can you confirm?

**Reply by brillout** [reply]
> (I'll improve the error message in a moment.)

Done https://github.com/brillout/vite-plugin-ssr/commit/7ffb37300f8075d13145fa12b3b5255090dcea3b.

**Comment by advancedtw** [comment]
thanks I managed to make the clouflare version work by changing a couple of things. the first is the handler now returns 
```js
  new Response(httpResponse.getReadableWebStream(), {status: httpResponse.statusCode,});
```
prior to that change i was getting 
```
✘ [ERROR] Uncaught (in promise) TypeError: This WritableStream has been closed.
```
I also had to install 2 libraries before I could get any kind of response "@brillout/json-serializer" and "scheduler" but then saw that I had preserveSymlinks in my vite config so I removed that and things started to work normally wothout these 2 libs.

for local node (btw i'm running v18.16) I still can't make it work. I set webstream to true in renderToStream and now I get a different error.

```
Error: [react-streaming@0.3.12][Wrong Usage] Your environment seems broken. (Could not import `renderToReadableStream` from `react-dom/server`). Create a new GitHub issue at https://github.com/brillout/react-streaming to discuss a solution.
```

**Reply by brillout** [reply]
> Error: [react-streaming@0.3.12][Wrong Usage] Your environment seems broken. (Could not import `renderToReadableStream` from `react-dom/server`). Create a new GitHub issue at https://github.com/brillout/react-streaming to discuss a solution.

~~That's quite surprising. You shouldn't see that if you use React 18. Maybe wrangler 3 picks the wrong `package.json#exports` of React?~~ Or do you use `renderToStream({ webStream: true })`?

Btw. I'm curious: what is it you're building? Why are you using `react-streaming`?

**Comment by advancedtw** [comment]
> Or do you use renderToStream({ webStream: true })?

yes but that was just me trying to force a webstream and see how node would handle that. 

I also wanted to clarify that I successfully created a demo working with streaming on both node and cloudflare but  I'm stuck with the hattip integration.  so vps and react-streaming are working as intended.

> Btw. I'm curious: what is it you're building? Why are you using react-streaming?

nothing fancy just a dashboard kinda app, I'm converting a solid spa to react ssr  and some of the pages pull data from slow and remote sources (different continent !). I'm hoping that by using streaming the user will be able to interact with the app before waiting for all the components to be loaded. 
I could be using next instead but with the recent release they broke every ui framework.

**Reply by brillout** [reply]
I think the bottom line is that you need a Web Stream in your Node.js dev environment. Can `renderToReadableStream()` be imported from React in a Node.js environment?

**Reply by brillout** [reply]
https://github.com/brillout/react-streaming/commit/22494a17177c6f3a35eb7bbd25d791bba18f74c0 released in `react-streaming@0.3.13`. Let me know if that works.

**Reply by advancedtw** [reply]
yes that did the trick. using  ```webStream: true``` and ```react-streaming@0.3.1``` and now I can use node.js however the new update breaks couldflare integration.
```
✘ [ERROR] Could not resolve "stream"

    node_modules/.pnpm/react-dom@18.2.0_react@18.2.0/node_modules/react-dom/cjs/react-dom-server-legacy.node.development.js:18:21:
      18 │ var stream = require('stream');

✘ [ERROR] Could not resolve "util"

    node_modules/.pnpm/react-dom@18.2.0_react@18.2.0/node_modules/react-dom/cjs/react-dom-server.node.development.js:18:19:
      18 │ var util = require('util');
         ╵                    ~~~~~~
```
### possible workaround
- patch these two modules using ```node_compat = true ``` in wrangler.toml
- use [bundler-cloudflare-workers](https://github.com/hattipjs/hattip/tree/main/packages/bundler/bundler-cloudflare-workers) and leave the wrangler.toml unchanged

**Reply by brillout** [reply, marked as answer]
https://github.com/brillout/react-streaming/commit/02449a76343ce3318622a5578b4a7a97e8039b68 released in `0.3.14`.

Let me know if that works.

**Reply by advancedtw** [reply]
yes this version works perfectly with cf and node. now time to try the other integration (vercel,netifly, lagon ...).
i may create a pr with an example once I'm done tinkering. i still have to integrate telefunc and mixed rendering (static and streaming).

**Reply by brillout** [reply]
> i may create [...] an example

Looking forward to it. Doesn't need to be a PR (a standalone example at `github.com/advancedtw` will suffice and I'll add links to it in the docs).

Also in case you think you can convince your company: https://github.com/sponsors/brillout.
